rollup.config.js

```diff
@@ -106,6 +107,7 @@ export default async function ({ watch }) {
           plugins: [
             { resolveFileUrl },
             OMT({ loader: await omtLoaderPromise }),
+            importMetaAssets(),
```

Comment:
Hmm what is "static build" / what is the scenario where it wouldn't work? I've included it in the same place as `OMT` because conceptually both handle similar code paths (OMT handles `new Worker(new URL(..., import.meta.url))` already, and this plugin handles the rest of `new URL(..., import.meta.url)` not covered by OMT).

Reply:
`src/shared/prerendered-app/Intro/index.tsx` for example. This file is used to generate our initial HTML. If I change:

```js
import deviceScreenIcon from 'url:./imgs/demos/icon-demo-device-screen.jpg';
```

to

```js
const deviceScreenIcon = new URL('./imgs/demos/icon-demo-device-screen.jpg', import.meta.url).href;
```

(and work around the typescript issues) it works fine on the client, but the initial render contains a `file:` URL.

Reply:
Ah, interesting. How come we didn't run into that with OMT? E.g. I already rely on it bundling the `new Worker(...)` from wasm-bindgen-rayon, I guess that one also results in `file:` URL on `dev` branch right now.

Reply:
`new Worker` will only appear in JS aimed at the client

Reply:
Ok, sorry, I'm very confused again. What is "client" in this scenario? Browser? If so, what else can JS be aimed at?

Reply:
Client is the browser, yeah. Our build system is also JavaScript, so the code that runs to generate the static html is JavaScript, but not running in the browser (running in Node).

We have three root component directories:

- `static-build` - these components only execute as part of the build
- `client` - these components only execute in the browser
- `shared` - these components execute in both

Right now the URL pattern only works in the client code. This is a bit of a gotcha with the shared code, since it's easy to miss that it isn't working.

It might just be a matter of moving it to the common plugins.

Reply:
Oh ok, I see, so this is not something that's happening now (that's why I didn't see the issue), but something that might happen if we convert more of those URL imports to this pattern. Yeah, makes sense to move to common plugins to prevent this issue.

Reply:
Hm, I actually don't see that happening even now in static HTML. If I add

```js
//@ts-ignore
console.log(new URL('./imgs/demos/icon-demo-large-photo.jpg', import.meta.url).href);
```

to `src/shared/prerendered-app/Intro/index.tsx` as you suggested, then the `.tmp/build/static/index.html` output contains:

```
console.log(new URL("/c/icon-demo-large-photo-18da387a.jpg", module.uri).href);
```

Which seems fine and isn't a `file:` URL.

Did you mean a different output?

Reply:
(correspondingly, moving to `commonPlugins` doesn't seem to make a difference)

Reply:
Here's a demo https://deploy-preview-1038--squoosh.netlify.app/ (PR: https://github.com/GoogleChromeLabs/squoosh/pull/1038).

If you view-source you'll see:

```html
<img class="_demo-icon_5wn1n_222" src="file:///opt/build/repo/.tmp/build/shared/prerendered-app/Intro/imgs/demos/icon-demo-device-screen.jpg" alt="Device screen" />
```

If I move the plugin to `commonPlugins` then I get:

```html
<img class="_demo-icon_5wn1n_222" src="file:///c/icon-demo-device-screen-5d52d8b9.jpg" alt="Device screen" />
```

…which is closer, but still not right.

Reply:
Interesting. Which folder is that HTML generated in? I only did "search in files" for "file:" and couldn't find this one.

Reply:
Oh interesting, I can repro with your change. I wonder why it inlines `file:` when it's in HTML but correct URL when it's something like my `console.log`...

Reply:
I suppose it's because Rollup generates stuff like

```js
const deviceScreenIcon = new URL('./imgs/demos/icon-demo-device-screen.jpg', (typeof document === 'undefined' ? new (require('u' + 'rl').URL)('file:' + __filename).href : (document.currentScript && document.currentScript.src || new URL('shared/prerendered-app/Intro/index.js', document.baseURI).href))).href;
```

so in browser where `document` is defined, it works fine, but in static render it falls back to the `file:` stuff.

I'll look into it a bit more.

Reply:
Where were you looking for the log? In the console of the build, or in the browser?

Reply:
Ah it is. I guess I just wasn't looking for that log line, I see it now.

Reply:
I guess we could fix that bit in another PR. It's less important than the wasm bloat.

Reply:
Let me take another look and if not, yeah, let's postpone.

At least now I see what's happening here. The `new URL(..., import.meta.url)` is meant to transform in a way where 2nd arg becomes a valid fully-specified URL. However, there is no such URL Rollup can substitute in Node.js since it doesn't know the domain application will end up with, and substituting with just relative URL like `/` would make `new URL` invocation invalid.

I'm looking at options of the plugin to see if we can work around this issue somehow...

Reply:
> I'm looking at options of the plugin to see if we can work around this issue somehow...

Yeah it doesn't seem like plugin supports that.

Reply:
Okay so I can't think of a good solution to support this right now, but at least I've added a custom `resolveImportMeta` handler to static build that prevents usages like the one in your example and throws a build-time error for now.

It still accepts all the usages in this PR, so it seems like a good balanced solution until we find a way to support `new URL` in static HTML too. WDYT?

Reply:
Sounds good